### PR TITLE
codecov 

### PR DESCRIPTION
**Pull request by rienafairefr**
codecov is run inside the tests docker container

we use tox as ENTRYPOINT in the tests docker container

tested working on my fork https://codecov.io/gh/rienafairefr/iot-lab-gateway/branch/codecov

There is still some issue about getting the coverage for the control_node_serial (the C code that interfaces with the control node) probably to do with relative/absolute paths

### CODE REVIEW

**Comment by fsaintma** [comment]
Hi when I have tested CI integration with install-lib I realized that I cannot launch 
control_node_serial_interface tests because of binary gcov and when I fixed it the coverage results seemed to be empty.

```
root@leonardo-1:/tmp/iot-lab-gateway# export GCOV=arm-poky-linux-gnueabi-gcov
root@leonardo-1:/tmp/iot-lab-gateway# tox -e control_node_serial
....
  GCOV  results/c_code_coverage.xml
  GCOV  STDOUT
------------------------------------------------------------------------------
                           GCC Code Coverage Report
Directory: ../../
------------------------------------------------------------------------------
File                                       Lines    Exec  Cover   Missing
------------------------------------------------------------------------------
------------------------------------------------------------------------------
TOTAL                                          0       0    --%
------------------------------------------------------------------------------
  GCOV  STDOUT
------------------------------------------------------------------------------
                           GCC Code Coverage Report
Directory: ../../
------------------------------------------------------------------------------
File                                    Branches   Taken  Cover   Missing
------------------------------------------------------------------------------
------------------------------------------------------------------------------
TOTAL                                          0       0    --%
------------------------------------------------------------------------------
make: Leaving directory '/var/volatile/tmp/iot-lab-gateway/control_node_serial'
__________________________________________ summary ___________________________________________
  control_node_serial: commands succeeded
```

So I try to understand and I modify control_node_serial/tests/Makefile and remove -e option of the following line to print coverage results:
GCOVRFLAGS  = --root '../../' -e '.*/tests' 

But I didn't really test and validate this modification :) But maybe it's a start of fix ...

**Comment by rienafairefr** [comment]
I've reworked the code (mainly ignoring the coverage inside control_node_serial/tests/utils), and it seems to work ok, including the coverage of the C code inside control_node_serial

File: tests/Dockerfile [hunk 1]
```diff
@@ -21,7 +21,7 @@ FROM iot-lab-gateway
 
 # Update
 RUN apt-get update && \
-    apt-get install -y python-pip gdb
+    apt-get install -y python-pip valgrind gdb
```
Comment (aabadie) [inline tests/Dockerfile:24]: This looks unrelated

File: tox.ini [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = copying, test, code_check
+envlist = copying, test, code_check, control_node_serial
```
Comment (aabadie) [inline tox.ini:2]: It is related ?

File: codecov.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+ignore:
+  - "control_node_serial/tests/utils"
```
Comment (aabadie) [inline codecov.yml]: Missing newline at end of file

File: .travis.yml [hunk 4]
```diff
@@ -13,18 +13,24 @@ before_install:
...
-    docker run -v gateway_code:/home/iot-lab-gateway/gateway_code
-    -v $CWD/tests_utils/cfg_dir:/var/local/config
+    docker run
+    -v $PWD/tests_utils/cfg_dir:/var/local/config
+    -v $PWD:/home/iot-lab-gateway
     -e BOARD_TYPE=m3
     -e CONTROL_NODE_TYPE=no
     -e HOSTNAME=m3-00 
     -e IOTLAB_GATEWAY_NO_INTEGRATION_TESTS
+    $ci_env
```
Comment (aabadie) [inline .travis.yml]: Is this still needed ?

**Review by aabadie — CHANGES_REQUESTED** [review]
Just some comments about changes that seem unrelated are not needed.

File: tests/Dockerfile (hunk 1 above)
Comment (rienafairefr) [reply tests/Dockerfile:24]: This is because we're doing the coverage for the control_node_serial as well, and valgrind is a dependency for that

File: tox.ini (hunk 2 above)
Comment (rienafairefr) [reply tox.ini:2]: Yes, I think we should test the coverage of the C code as well

File: .travis.yml (hunk 4 above)
Comment (rienafairefr) [reply .travis.yml]: Yes, basically we're running codecov inside the docker container, it needs to know some things about the CI environment which are in TRAVIS_* environment variables, this is a shortcut to pass them all at once, see https://github.com/codecov/support/wiki/Testing-with-Docker

File: codecov.yml (hunk 3 above)
Comment (rienafairefr) [reply codecov.yml]: OK, fixed

File: .travis.yml (hunk 4 above)
Comment (aabadie) [reply .travis.yml]: ok so what is the point of the call to upload_coverage on success ?

File: tests/Dockerfile (hunk 1 above)
Comment (rienafairefr) [reply tests/Dockerfile:24]: I could separate in 2 PRs (one to test control_node_serial, one for codecov), is that necessary ?

File: .travis.yml (hunk 4 above)
Comment (rienafairefr) [reply .travis.yml]: It's uploading the coverage to codecov.io when the tests are successful

File: tests/Dockerfile (hunk 1 above)
Comment (aabadie) [reply tests/Dockerfile:24]: > I've reworked the code (mainly ignoring the coverage inside control_node_serial/tests/utils)

You are saying 2 contradictory things. I'm a bit lost.

No need for 2 PRs, just good reasons

File: .travis.yml (hunk 4 above)
Comment (aabadie) [reply .travis.yml]: Yes, using codecov. I think the codecov environment variables (from $ci_env) are only useful in this case.

File: tests/Dockerfile (hunk 1 above)
Comment (rienafairefr) [reply tests/Dockerfile:24]: before the code I added yesterday, we had coverage in codecov.io just fine, but the coverage for control_node_serial was artifically low because it was not ignoring the control_node_serial/tests/utils folder (which contains some binary files). Sorry if I'm not clear ^^

File: .travis.yml (hunk 4 above)
Comment (rienafairefr) [reply .travis.yml]: Oh now I see what you mean. You're right, this $ci_env shouldnt be needed in the docker run in script.

File: codecov.yml [hunk 5]
```diff
@@ -0,0 +1,2 @@
+ignore:
```
Comment (aabadie) [inline codecov.yml:1]: Maybe you can also disable the comment in the PR: `comment: false`

File: tests/Dockerfile (hunk 1 above)
Comment (aabadie) [reply tests/Dockerfile:24]: That's fine, I understand now.

File: codecov.yml (hunk 5 above)
Comment (rienafairefr) [reply codecov.yml:1]: True, I'll add it. Thanks

**Comment by rienafairefr** [comment]
Full squash before merge ?

**Comment by aabadie** [comment]
> Full squash before merge ?

That would be better, thanks !

**Comment by rienafairefr** [comment]
OK, squashed an tested, I'm merging
